src/clj/rems/application/commands.clj

```diff
@@ -492,14 +491,47 @@
        :application/member (:member cmd)
        :invitation/token (secure-token)}))
 
+(defmethod command-handler :application.command/invite-actor
+  [cmd _application {:keys [secure-token]}]
+  (ok {:event/type :application.event/actor-invited
+       :application/actor (:invitee cmd)
+       :invitation/role (:role cmd)
+       :invitation/token (secure-token)}))
+
+(defn- valid-member-invitation-token? [application token]
+  (contains? (:application/invitation-tokens application) token))
+
+(defn- get-actor-invitation-token [application token]
+  (get-in application [:application/actor-invitations token]))
+
 (defmethod command-handler :application.command/accept-invitation
   [cmd application _injections]
-  (or (already-member-error application (:actor cmd))
-      (invitation-token-error application (:token cmd))
+  (let [token (:token cmd)
+        actor-invitation (get-actor-invitation-token application token)]
+    (cond
+      (valid-member-invitation-token? application token)
+      (or (already-member-error application (:actor cmd))
+          (ok-with-data {:application-id (:application-id cmd)}
+                        [{:event/type :application.event/member-joined
+                          :application/id (:application-id cmd)
+                          :invitation/token (:token cmd)}]))
+      actor-invitation
+      ;; TODO it's nice to separate the concerns by emitting two
+      ;; events here, but the downsides are:
+      ;; - potentially confusing event log in UI
+      ;; - extra email sent to reviewer ("foo has requested your review on bar") after accepting invitation
```

Comment:
Commit 2049a3651bc831f37433d5a046489b022fff42e8 has an implementation where the actor-joined event does everything, but that led to a bit of code duplication. I prefer this split, but the log & email issues mentioned here need to be resolved somehow. Some sort of `:silent true` flag for the `review-requested` event? Split `review-requested` event into two events (one sends notifications & shows up in log, one does the model updates)?